### PR TITLE
Vircadia client library messaging APIs.

### PR DESCRIPTION
**Pull request by namark**
Builds upon https://github.com/vircadia/vircadia/pull/1623

- [x] API implementation
- [x] API unit tests
- [x] API documentation

The unit test will attempt to connect to localhost and send a message in "Chat" channel in appropriate JSON format, so it should appear in the Domain tab of the native interface chat, and otherwise it will test just a few error cases.

### CODE REVIEW

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-a6ce2f2-_425ca1f3.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-a6ce2f2-.tar_425ca1f3.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-a6ce2f2-.tar_425ca1f3.Z

**build (windows-2019, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-a6ce2f2-_425ca1f3.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-10.15, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-Interface-PR1630-a944d02-_6cc1599f.dmg

**build (macOS-10.15, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-a944d02-_6cc1599f.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-a944d02-.tar_6cc1599f.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-a944d02-.tar_6cc1599f.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-a944d02-_6cc1599f.sh

**build (windows-2019, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-a944d02-_6cc1599f.exe

File: tests/ktx/src/KtxTests.cpp [hunk 1]
```diff
@@ -12,6 +12,10 @@
 
 #include <QtTest/QtTest>
 
+#if 0
```
Comment (DelGuapo) [inline tests/ktx/src/KtxTests.cpp:15]: ?

**Review by DelGuapo — APPROVED** [review]
Very impressed... I think you left some scrap test code in tests/ktx/src/KtxTests.cpp.

Looks good - can proceed.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-10.15, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-1b82ed6-_a9370b4d.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-1b82ed6-.tar_a9370b4d.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-1b82ed6-.tar_a9370b4d.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-1b82ed6-_a9370b4d.sh

**build (macOS-10.15, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-Interface-PR1630-1b82ed6-_a9370b4d.dmg

**build (windows-2019, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-1b82ed6-_a9370b4d.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-10.15, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-3732b4c-_df806442.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-3732b4c-_df806442.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-3732b4c-.tar_df806442.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-3732b4c-.tar_df806442.gz

**build (macOS-10.15, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-Interface-PR1630-3732b4c-_df806442.dmg

**build (windows-2019, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-3732b4c-_df806442.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-10.15, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-Interface-PR1630-9eb95ab-_158f8faf.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-9eb95ab-.tar_158f8faf.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-9eb95ab-.tar_158f8faf.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-9eb95ab-_158f8faf.sh

**Review by DelGuapo — APPROVED** [review]

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-10.15, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-Interface-PR1630-3d73ec4-_ddd950d1.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-3d73ec4-.tar_ddd950d1.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-3d73ec4-.tar_ddd950d1.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-3d73ec4-_ddd950d1.sh

**build (macOS-10.15, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-3d73ec4-_ddd950d1.dmg

**build (windows-2019, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-3d73ec4-_ddd950d1.exe

File: libraries/vircadia-client/src/error.cpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
+#include "error.h"
```
Comment (ctrlaltdavid) [inline libraries/vircadia-client/src/error.cpp:1]: Copyright & license header missing.

File: libraries/vircadia-client/src/error.h [hunk 3]
```diff
@@ -0,0 +1,71 @@
+//
+//  error.h
+//  libraries/client/src
```
Comment (ctrlaltdavid) [inline libraries/vircadia-client/src/error.h:3]:
```suggestion
//  libraries/vircadia-client/src
```
Here and elsewhere.

File: libraries/vircadia-client/src/error.h [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+#include "common.h"
+
+/// @brief Only one context can exist at a time.
+///
+/// A limitation in the current library implantation, this error code is
+/// returned if user attempts to create the second context.
+///
+/// @return -1
+VIRCADIA_CLIENT_DYN_API int vircadia_error_context_exists();
```
Comment (ctrlaltdavid) [inline libraries/vircadia-client/src/error.h:25]: There's a question over the use of "_"s in names.

File: libraries/vircadia-client/src/internal/Context.h [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+    private:
+        std::thread appThraed {};
+        std::atomic<QCoreApplication*> app {};
+        std::vector<NodeData> nodes {};
+        std::promise<void> qtInitialized {};
+
+
+        struct MessagesContext {
+            bool enabled {false};
```
Comment (ctrlaltdavid) [inline libraries/vircadia-client/src/internal/Context.h:106]:
```suggestion
            bool enabled { false };
```

File: libraries/vircadia-client/src/internal/Context.h [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+        int argc;
+        char argvData[18];
+        char* argv;
+    };
+
+    extern std::list<Context> contexts;
+
+    int vircadiaContextValid(int id);
+    int vircadiaContextReady(int id);
```
Comment (ctrlaltdavid) [inline libraries/vircadia-client/src/internal/Context.h:122]: Should these return a bool?
Better names would probably start with "is" per naming conventions.

File: libraries/vircadia-client/src/internal/Context.h [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+        void connect(const QString& address) const;
+
+        void enableMessages(std::bitset<8> type);
+        void updateMessages(std::bitset<8> type);
+        void clearMessages(std::bitset<8> type);
+        void subscribeMessages(QString channel) const;
+        void unsubscribeMessages(QString channel) const;
+        const std::vector<MessageData>& getMessages(std::bitset<8> type) const;
+        bool getMessagesEnabled(std::bitset<8> type) const;
```
Comment (ctrlaltdavid) [inline libraries/vircadia-client/src/internal/Context.h:95]:
```suggestion
        bool isMessagesEnabled(std::bitset<8> type) const;
```

File: libraries/vircadia-client/src/internal/Error.h [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+            return *error;
+        }
+    } else {
+        return std::invoke(std::forward<F>(f), result);
+    }
+};
+
+template <typename Container>
+int indexValid(const Container& container, int index, ErrorCode code) {
+    if ( index < 0 || index >= static_cast<int>(container.size()) ) {
```
Comment (ctrlaltdavid) [inline libraries/vircadia-client/src/internal/Error.h:82]:
```suggestion
    if (index < 0 || index >= static_cast<int>(container.size())) {
```

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]
The naming of files in /libraries/vircadia-client appears to be contrary to Vircadia conventions: starting with a lower case letter (convention is uppercase) and using "_"s in names (convention is upper camel case).

**Comment by namark** [comment]
> The naming of files in /libraries/vircadia-client appears to be contrary to Vircadia conventions: starting with a lower case letter (convention is uppercase) and using "_"s in names (convention is upper camel case).

This is similar to function names, wanted to make the API interface stand out and be more C like, these again should not be included anywhere in the repo, other than the unit tests. The `src/internal` folder has the C++ implementation details where I tried to follow our coding style, so there is some blending of styles going on between the two, but yeah in general I try to stick to C style in API interface (what will be exposed to the potential users) and our usual style in the implementation details.

File: libraries/vircadia-client/src/error.h (hunk 4 above)
Comment (namark) [reply libraries/vircadia-client/src/error.h:25]: Tried to clarified in #1623

File: libraries/vircadia-client/src/internal/Context.h (hunk 6 above)
Comment (namark) [reply libraries/vircadia-client/src/internal/Context.h:122]: These are more like error checks, than boolean getters, they return 0 if there is no error, or a specific negative error code. Other API functions forward error codes that they get from these. This makes the overall error reporting code a bit easier to work with, and also results in the error codes being specified next to (or close to) the actual check in the code, so it's easier to see which conditions correspond to which error codes.

These are in the vircadia::client namespace though, and don't need the vircadia prefix (I initially wrote them as C API functions then moved them here) so I'll rename them to `checkContextValid` and `checkContextReady`.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-10.15, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-Interface-PR1630-7441a1f-_f5fde75c.dmg

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-10.15, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-Interface-PR1630-4b9fe60-_a30d3b08.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-4b9fe60-.tar_a30d3b08.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-4b9fe60-.tar_a30d3b08.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-4b9fe60-_a30d3b08.sh

**build (macOS-10.15, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-4b9fe60-_a30d3b08.dmg

**build (windows-2019, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1630/Vircadia-PR1630-4b9fe60-_a30d3b08.exe

**Review by ctrlaltdavid — APPROVED** [review]

**Pull request merged**